### PR TITLE
fix(server): AsyncWaiters sends a response back to the waiting internal server

### PR DESCRIPTION
**Pull request by eduwercamacaro**

### CODE REVIEW

File: schemas/internal/interactive_query.proto [hunk 1]
```diff
@@ -21,12 +21,18 @@ message GetObjectResponse {
...
+    int32 partition = 2;
 }
 
 message WaitForCommandResponse {
     string command_id = 1;
     google.protobuf.Timestamp result_time = 2;
-    bytes result = 3;
+    message PartitionMigratedResponse {
+    }
+    oneof yield {
```
Comment (coltmcnealy-lh) [inline schemas/internal/interactive_query.proto:32]: `pedro`

File: server/src/main/java/io/littlehorse/server/monitoring/HealthService.java
Comment (coltmcnealy-lh) [inline server/src/main/java/io/littlehorse/server/monitoring/HealthService.java:1]: This is not a request to change this now, but the fact that we are editing the `HealthService` smells a bit wrong. I think we should probably have a separate "Rebalance Listener" class that takes in the `BackendInternalComms` and the `HealthService` as dependencies.

But I don't think it's a huge problem for now, and we can fix it later. It's not customer facing and wont' impact performance.

File: server/src/main/java/io/littlehorse/server/streams/util/AsyncWaiters.java [hunk 2]
```diff
@@ -53,8 +56,9 @@ public AsyncWaiters() {
                 .start();
     }
 
-    public void registerObserverWaitingForCommand(String commandId, StreamObserver<WaitForCommandResponse> observer) {
-        CommandWaiter tmp = new CommandWaiter(commandId);
+    public void registerObserverWaitingForCommand(
+            String commandId, int partition, StreamObserver<WaitForCommandResponse> observer) {
```
Comment (coltmcnealy-lh) [inline server/src/main/java/io/littlehorse/server/streams/util/AsyncWaiters.java:60]: Nice, partition-awareness :smile: Somebody knows Kafka really well

File: server/src/main/java/io/littlehorse/server/streams/util/AsyncWaiters.java [hunk 3]
```diff
@@ -122,6 +126,15 @@ public void registerObserverWaitingForWorkflowEvent(
         }
     }
 
+    public void handleRebalance(Set<TaskId> assignedTasks) {
+        Set<Integer> assignedPartitions =
+                assignedTasks.stream().map(TaskId::partition).collect(Collectors.toSet());
+
+        commandWaiters.values().stream()
+                .filter(commandWaiter -> !assignedPartitions.contains(commandWaiter.getCommandPartition()))
+                .forEach(migratedCommand -> migratedCommand.handleMigration());
```
Comment (coltmcnealy-lh) [inline server/src/main/java/io/littlehorse/server/streams/util/AsyncWaiters.java:135]: Small nit pick: this should be `migratedCommandWaiter` not `migratedCommand`

File: server/src/main/java/io/littlehorse/server/streams/util/POSTStreamObserver.java [hunk 4]
```diff
@@ -43,6 +54,10 @@ public void onCompleted() {
 
     @Override
     public void onNext(WaitForCommandResponse reply) {
-        ctx.onNext(buildRespFromBytes(reply.getResult()));
+        if (reply.hasResult()) {
+            ctx.onNext(buildRespFromBytes(reply.getResult()));
+        } else if (reply.hasPartitionMigratedResponse()) {
+            internalComms.waitForCommand(command, this);
```
Comment (coltmcnealy-lh) [inline server/src/main/java/io/littlehorse/server/streams/util/POSTStreamObserver.java:60]: Honestly, I'm surprised it was so simple. Nice!

**Review by coltmcnealy-lh — APPROVED** [review]
This should be a massive improvement to reliability.

**Review by Snarr — APPROVED** [review]

**Pull request merged**